From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: timoreo <timo.oreo34@gmail.com>
Date: Wed, 7 Jul 2021 18:01:50 +0200
Subject: [PATCH] Removed Dupe Patches


diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index 33f41de3aabaee37dbd92ad17d8dc65a60fef427..c022ae11f911e81c678b2a9648d8b695827ec005 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -2430,7 +2430,7 @@ public class ServerPlayer extends Player {
 
     @Override
     public boolean isImmobile() {
-        return super.isImmobile() || frozen || (this.connection != null && this.connection.isDisconnected()); // Paper // Purpur
+       return super.isImmobile() || frozen || !this.getBukkitEntity().isOnline(); // Purpur
     }
 
     // Purpur start
diff --git a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index 0d3e73240e5d363427379f50ecb163482c6f1bfc..65ff946dbfd451f82d1493ccf7ccfaf412e6d540 100644
--- a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -3144,7 +3144,7 @@ public class ServerGamePacketListenerImpl implements ServerPlayerConnection, Ser
     // Paper end
 
     public final boolean isDisconnected() {
-        return (!this.player.joining && !this.connection.isConnected()) || this.processedDisconnect; // Paper
+        return !this.player.joining && !this.connection.isConnected();
     }
     // CraftBukkit end
 
diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index a75b0a7d8bbb6ed0871865c70b5a5256a4c61cbe..033c0958b08cefffb4936cde0da991f6ba5dc441 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -2214,12 +2214,11 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource, n
         } else {
             // CraftBukkit start - Capture drops for death event
             if (this instanceof net.minecraft.world.entity.LivingEntity && !((net.minecraft.world.entity.LivingEntity) this).forceDrops) {
-                ((net.minecraft.world.entity.LivingEntity) this).drops.add(org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(stack)); // Paper - mirror so we can destroy it later
+                ((net.minecraft.world.entity.LivingEntity) this).drops.add(org.bukkit.craftbukkit.inventory.CraftItemStack.asBukkitCopy(stack));
                 return null;
             }
             // CraftBukkit end
-            ItemEntity entityitem = new ItemEntity(this.level, this.getX(), this.getY() + (double) yOffset, this.getZ(), stack.copy()); // Paper - clone so we can destroy original
-            stack.setCount(0); // Paper - destroy this item - if this ever leaks due to game bugs, ensure it doesn't dupe
+            ItemEntity entityitem = new ItemEntity(this.level, this.getX(), this.getY() + (double) yOffset, this.getZ(), stack);
 
             entityitem.setDefaultPickUpDelay();
             // CraftBukkit start
@@ -2979,12 +2978,6 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource, n
     @Nullable
     public Entity teleportTo(ServerLevel worldserver, BlockPos location) {
         // CraftBukkit end
-        // Paper start - fix bad state entities causing dupes
-        if (!isAlive() || !valid) {
-            LOGGER.warn("Illegal Entity Teleport " + this + " to " + worldserver + ":" + location, new Throwable());
-            return null;
-        }
-        // Paper end
         if (this.level instanceof ServerLevel && !this.isRemoved()) {
             this.level.getProfiler().push("changeDimension");
             // CraftBukkit start
@@ -3005,11 +2998,6 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource, n
                 // CraftBukkit end
 
                 this.level.getProfiler().popPush("reloading");
-                // Paper start - Change lead drop timing to prevent dupe
-                if (this instanceof Mob) {
-                    ((Mob) this).dropLeash(true, true); // Paper drop lead
-                }
-                // Paper end
                 Entity entity = this.getType().create((Level) worldserver);
 
                 if (entity != null) {
@@ -3023,6 +3011,9 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource, n
                     // CraftBukkit start - Forward the CraftEntity to the new entity
                     this.getBukkitEntity().setHandle(entity);
                     entity.bukkitEntity = this.getBukkitEntity();
+                    if (this instanceof Mob) {
+                        ((Mob) this).dropLeash(true, false); // Unleash to prevent duping of leads.
+                    }
                     // CraftBukkit end
                 }
 
@@ -3147,7 +3138,7 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource, n
     }
 
     public boolean canChangeDimensions() {
-        return isAlive() && valid && (level.purpurConfig.entitiesCanUsePortals || this instanceof ServerPlayer); // Paper // Purpur
+        return (level.purpurConfig.entitiesCanUsePortals || this instanceof ServerPlayer); // Paper // Purpur
     }
 
     public float getBlockExplosionResistance(Explosion explosion, BlockGetter world, BlockPos pos, BlockState blockState, FluidState fluidState, float max) {
diff --git a/src/main/java/net/minecraft/world/entity/decoration/ArmorStand.java b/src/main/java/net/minecraft/world/entity/decoration/ArmorStand.java
index 6d08c8c31a32ea38f06410fbaddf19b95bec7c6f..13ec8823b2cbb78608b84761ab40ec87a4409965 100644
--- a/src/main/java/net/minecraft/world/entity/decoration/ArmorStand.java
+++ b/src/main/java/net/minecraft/world/entity/decoration/ArmorStand.java
@@ -655,7 +655,7 @@ public class ArmorStand extends LivingEntity {
         for (i = 0; i < this.handItems.size(); ++i) {
             itemstack = (ItemStack) this.handItems.get(i);
             if (!itemstack.isEmpty()) {
-                drops.add(org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(itemstack)); // CraftBukkit - add to drops // Paper - mirror so we can destroy it later - though this call site was safe
+                drops.add(org.bukkit.craftbukkit.inventory.CraftItemStack.asBukkitCopy(itemstack)); // CraftBukkit - add to drops
                 this.handItems.set(i, ItemStack.EMPTY);
             }
         }
@@ -663,7 +663,7 @@ public class ArmorStand extends LivingEntity {
         for (i = 0; i < this.armorItems.size(); ++i) {
             itemstack = (ItemStack) this.armorItems.get(i);
             if (!itemstack.isEmpty()) {
-                drops.add(org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(itemstack)); // CraftBukkit - add to drops // Paper - mirror so we can destroy it later - though this call site was safe
+                drops.add(org.bukkit.craftbukkit.inventory.CraftItemStack.asBukkitCopy(itemstack)); // CraftBukkit - add to drops
                 this.armorItems.set(i, ItemStack.EMPTY);
             }
         }
diff --git a/src/main/java/net/minecraft/world/entity/player/Player.java b/src/main/java/net/minecraft/world/entity/player/Player.java
index f8cf130aee9bbdb62b23809b58eece968a01d3da..46e9425f3aae8dfa066b1d5012008266715b8b79 100644
--- a/src/main/java/net/minecraft/world/entity/player/Player.java
+++ b/src/main/java/net/minecraft/world/entity/player/Player.java
@@ -746,11 +746,6 @@ public abstract class Player extends LivingEntity {
             }
 
             double d0 = this.getEyeY() - 0.30000001192092896D;
-            // Paper start
-            ItemStack tmp = stack.copy();
-            stack.setCount(0);
-            stack = tmp;
-            // Paper end
             ItemEntity entityitem = new ItemEntity(this.level, this.getX(), d0, this.getZ(), stack);
 
             entityitem.setPickUpDelay(40);
diff --git a/src/main/java/net/minecraft/world/inventory/HorseInventoryMenu.java b/src/main/java/net/minecraft/world/inventory/HorseInventoryMenu.java
index ffc596d8db77aaeed43f79b895bf4a1c7baeeab2..f3e17f00f09feb4df4274b09bc7fa3c3022a5fbb 100644
--- a/src/main/java/net/minecraft/world/inventory/HorseInventoryMenu.java
+++ b/src/main/java/net/minecraft/world/inventory/HorseInventoryMenu.java
@@ -95,7 +95,7 @@ public class HorseInventoryMenu extends AbstractContainerMenu {
 
     @Override
     public boolean stillValid(Player player) {
-        return !this.horse.hasInventoryChanged(this.horseContainer) && this.horseContainer.stillValid(player) && (this.horse.isAlive() && this.horse.valid) && this.horse.distanceTo((Entity) player) < 8.0F; // Paper - Fix MC-161754
+        return !this.horse.hasInventoryChanged(this.horseContainer) && this.horseContainer.stillValid(player) && this.horse.isAlive() && this.horse.distanceTo((Entity) player) < 8.0F;
     }
 
     private boolean hasChest(AbstractHorse horse) {
diff --git a/src/main/java/net/minecraft/world/item/PotionItem.java b/src/main/java/net/minecraft/world/item/PotionItem.java
index 12a29323d99dcc7880fe3c7c9709a755d9cbf43e..9014bf545205504bee2d727399d39090ebe3d210 100644
--- a/src/main/java/net/minecraft/world/item/PotionItem.java
+++ b/src/main/java/net/minecraft/world/item/PotionItem.java
@@ -42,7 +42,6 @@ public class PotionItem extends Item {
             CriteriaTriggers.CONSUME_ITEM.trigger((ServerPlayer) entityhuman, stack);
         }
 
-        List<MobEffectInstance> instantLater = new java.util.ArrayList<>(); // Paper - Fix harming potion dupe
         if (!world.isClientSide) {
             List<MobEffectInstance> list = PotionUtils.getMobEffects(stack);
             Iterator iterator = list.iterator();
@@ -51,7 +50,7 @@ public class PotionItem extends Item {
                 MobEffectInstance mobeffect = (MobEffectInstance) iterator.next();
 
                 if (mobeffect.getEffect().isInstantenous()) {
-                    instantLater.add(mobeffect); // Paper - Fix harming potion dupe
+                    mobeffect.getEffect().applyInstantenousEffect(entityhuman, entityhuman, user, mobeffect.getAmplifier(), 1.0D);
                 } else {
                     user.addEffect(new MobEffectInstance(mobeffect), org.bukkit.event.entity.EntityPotionEffectEvent.Cause.POTION_DRINK); // CraftBukkit
                 }
@@ -65,18 +64,7 @@ public class PotionItem extends Item {
             }
         }
 
-        // Paper start - Fix harming potion dupe
-        for (MobEffectInstance mobeffect : instantLater) {
-            mobeffect.getEffect().applyInstantenousEffect(entityhuman, entityhuman, user, mobeffect.getAmplifier(), 1.0D);
-        }
-        // Paper end
         if (entityhuman == null || !entityhuman.getAbilities().instabuild) {
-            // Paper start - Fix harming potion dupe
-            if (user.getHealth() <= 0 && !user.level.getGameRules().getBoolean(net.minecraft.world.level.GameRules.RULE_KEEPINVENTORY)) {
-                user.spawnAtLocation(new ItemStack(Items.GLASS_BOTTLE), 0);
-                return ItemStack.EMPTY;
-            }
-            // Paper end
             if (stack.isEmpty()) {
                 return new ItemStack(Items.GLASS_BOTTLE);
             }
diff --git a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
index f60ef0605163ecf50d9f61a73a84146d214aaf19..a4a4121778a656f8778eb96710212e2fa71d83e4 100644
--- a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
+++ b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
@@ -862,8 +862,7 @@ public class CraftEventFactory {
         for (org.bukkit.inventory.ItemStack stack : event.getDrops()) {
             if (stack == null || stack.getType() == Material.AIR || stack.getAmount() == 0) continue;
 
-            world.dropItem(entity.getLocation(), stack); // Paper - note: dropItem already clones due to this being bukkit -> NMS
-            if (stack instanceof CraftItemStack) stack.setAmount(0); // Paper - destroy this item - if this ever leaks due to game bugs, ensure it doesn't dupe, but don't nuke bukkit stacks of manually added items
+            world.dropItem(entity.getLocation(), stack);
         }
 
         return event;
