From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: timoreo <timo.oreo34@gmail.com>
Date: Tue, 7 Dec 2021 07:17:24 +0100
Subject: [PATCH] Removed Dupe Patches


diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index 7b23535a680d2a8534dcb8dd87770f66fb982c13..5fe8d45cc5d61d2cb8f5ccf61319fab685682509 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -2388,7 +2388,7 @@ public class ServerPlayer extends Player {
 
     @Override
     public boolean isImmobile() {
-        return super.isImmobile() || (this.connection != null && this.connection.isDisconnected()); // Paper
+        return super.isImmobile();
     }
 
     @Override
diff --git a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index 348cecfe1c3ac1debe98e2fcc756c7e32a3187df..5d1614b6d850d01347153e8588d009eb3d1d62e3 100644
--- a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -3177,7 +3177,7 @@ public class ServerGamePacketListenerImpl implements ServerPlayerConnection, Ser
     // Paper end
 
     public final boolean isDisconnected() {
-        return (!this.player.joining && !this.connection.isConnected()) || this.processedDisconnect; // Paper
+        return !this.player.joining && !this.connection.isConnected();
     }
     // CraftBukkit end
 
diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index 821c34a48127802947e293c4599e0cdaea3c040e..19b2736eb6507a6f013200c66a500bef4f89b807 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -2411,12 +2411,11 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource, i
         } else {
             // CraftBukkit start - Capture drops for death event
             if (this instanceof net.minecraft.world.entity.LivingEntity && !((net.minecraft.world.entity.LivingEntity) this).forceDrops) {
-                ((net.minecraft.world.entity.LivingEntity) this).drops.add(org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(stack)); // Paper - mirror so we can destroy it later
+                ((net.minecraft.world.entity.LivingEntity) this).drops.add(org.bukkit.craftbukkit.inventory.CraftItemStack.asBukkitCopy(stack));
                 return null;
             }
             // CraftBukkit end
-            ItemEntity entityitem = new ItemEntity(this.level, this.getX(), this.getY() + (double) yOffset, this.getZ(), stack.copy()); // Paper - clone so we can destroy original
-            stack.setCount(0); // Paper - destroy this item - if this ever leaks due to game bugs, ensure it doesn't dupe
+            ItemEntity entityitem = new ItemEntity(this.level, this.getX(), this.getY() + (double) yOffset, this.getZ(), stack);
 
             entityitem.setDefaultPickUpDelay();
             // CraftBukkit start
@@ -3171,12 +3170,6 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource, i
     @Nullable
     public Entity teleportTo(ServerLevel worldserver, BlockPos location) {
         // CraftBukkit end
-        // Paper start - fix bad state entities causing dupes
-        if (!isAlive() || !valid) {
-            LOGGER.warn("Illegal Entity Teleport " + this + " to " + worldserver + ":" + location, new Throwable());
-            return null;
-        }
-        // Paper end
         if (this.level instanceof ServerLevel && !this.isRemoved()) {
             this.level.getProfiler().push("changeDimension");
             // CraftBukkit start
@@ -3216,11 +3209,6 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource, i
                 // CraftBukkit end
 
                 this.level.getProfiler().popPush("reloading");
-                // Paper start - Change lead drop timing to prevent dupe
-                if (this instanceof Mob) {
-                    ((Mob) this).dropLeash(true, true); // Paper drop lead
-                }
-                // Paper end
                 Entity entity = this.getType().create(worldserver);
 
                 if (entity != null) {
@@ -3234,6 +3222,9 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource, i
                     // CraftBukkit start - Forward the CraftEntity to the new entity
                     this.getBukkitEntity().setHandle(entity);
                     entity.bukkitEntity = this.getBukkitEntity();
+		    if(this instanceof Mob){
+			((Mob) this).dropLeash(true, false); // Unleash to prevent duping of leads.
+		    }
                     // CraftBukkit end
                 }
 
@@ -3354,7 +3345,7 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource, i
     }
 
     public boolean canChangeDimensions() {
-        return isAlive() && valid; // Paper
+        return true;
     }
 
     public float getBlockExplosionResistance(Explosion explosion, BlockGetter world, BlockPos pos, BlockState blockState, FluidState fluidState, float max) {
diff --git a/src/main/java/net/minecraft/world/entity/decoration/ArmorStand.java b/src/main/java/net/minecraft/world/entity/decoration/ArmorStand.java
index e5ef24d92de21c4c0e6a98e06985e52d47bfdce0..b89d60150c9b59ca2c3149b308423f4e40e7655d 100644
--- a/src/main/java/net/minecraft/world/entity/decoration/ArmorStand.java
+++ b/src/main/java/net/minecraft/world/entity/decoration/ArmorStand.java
@@ -610,7 +610,7 @@ public class ArmorStand extends LivingEntity {
         for (i = 0; i < this.handItems.size(); ++i) {
             itemstack = (ItemStack) this.handItems.get(i);
             if (!itemstack.isEmpty()) {
-                drops.add(org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(itemstack)); // CraftBukkit - add to drops // Paper - mirror so we can destroy it later - though this call site was safe
+                drops.add(org.bukkit.craftbukkit.inventory.CraftItemStack.asBukkitCopy(itemstack)); // CraftBukkit - add to drops
                 this.handItems.set(i, ItemStack.EMPTY);
             }
         }
@@ -618,7 +618,7 @@ public class ArmorStand extends LivingEntity {
         for (i = 0; i < this.armorItems.size(); ++i) {
             itemstack = (ItemStack) this.armorItems.get(i);
             if (!itemstack.isEmpty()) {
-                drops.add(org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(itemstack)); // CraftBukkit - add to drops // Paper - mirror so we can destroy it later - though this call site was safe
+                drops.add(org.bukkit.craftbukkit.inventory.CraftItemStack.asBukkitCopy(itemstack)); // CraftBukkit - add to drops
                 this.armorItems.set(i, ItemStack.EMPTY);
             }
         }
diff --git a/src/main/java/net/minecraft/world/entity/player/Player.java b/src/main/java/net/minecraft/world/entity/player/Player.java
index 2c00a766130a7f682fc6c4c74321e10637ca7932..7bfc17261ca1bcd811ea4f5ab995f03cbf7b005f 100644
--- a/src/main/java/net/minecraft/world/entity/player/Player.java
+++ b/src/main/java/net/minecraft/world/entity/player/Player.java
@@ -708,11 +708,6 @@ public abstract class Player extends LivingEntity {
             }
 
             double d0 = this.getEyeY() - 0.30000001192092896D;
-            // Paper start
-            ItemStack tmp = stack.copy();
-            stack.setCount(0);
-            stack = tmp;
-            // Paper end
             ItemEntity entityitem = new ItemEntity(this.level, this.getX(), d0, this.getZ(), stack);
 
             entityitem.setPickUpDelay(40);
diff --git a/src/main/java/net/minecraft/world/item/PotionItem.java b/src/main/java/net/minecraft/world/item/PotionItem.java
index 4b5c9da0e858016c1a3afcb6eaa1ff4cbf47739e..166f7483163a0cf62b000764d635ec2ee36a5036 100644
--- a/src/main/java/net/minecraft/world/item/PotionItem.java
+++ b/src/main/java/net/minecraft/world/item/PotionItem.java
@@ -41,7 +41,6 @@ public class PotionItem extends Item {
             CriteriaTriggers.CONSUME_ITEM.trigger((ServerPlayer) entityhuman, stack);
         }
 
-        List<MobEffectInstance> instantLater = new java.util.ArrayList<>(); // Paper - Fix harming potion dupe
         if (!world.isClientSide) {
             List<MobEffectInstance> list = PotionUtils.getMobEffects(stack);
             Iterator iterator = list.iterator();
@@ -50,7 +49,7 @@ public class PotionItem extends Item {
                 MobEffectInstance mobeffect = (MobEffectInstance) iterator.next();
 
                 if (mobeffect.getEffect().isInstantenous()) {
-                    instantLater.add(mobeffect); // Paper - Fix harming potion dupe
+                    mobeffect.getEffect().applyInstantenousEffect(entityhuman, entityhuman, user, mobeffect.getAmplifier(), 1.0D);
                 } else {
                     user.addEffect(new MobEffectInstance(mobeffect), org.bukkit.event.entity.EntityPotionEffectEvent.Cause.POTION_DRINK); // CraftBukkit
                 }
@@ -64,18 +63,7 @@ public class PotionItem extends Item {
             }
         }
 
-        // Paper start - Fix harming potion dupe
-        for (MobEffectInstance mobeffect : instantLater) {
-            mobeffect.getEffect().applyInstantenousEffect(entityhuman, entityhuman, user, mobeffect.getAmplifier(), 1.0D);
-        }
-        // Paper end
         if (entityhuman == null || !entityhuman.getAbilities().instabuild) {
-            // Paper start - Fix harming potion dupe
-            if (user.getHealth() <= 0 && !user.level.getGameRules().getBoolean(net.minecraft.world.level.GameRules.RULE_KEEPINVENTORY)) {
-                user.spawnAtLocation(new ItemStack(Items.GLASS_BOTTLE), 0);
-                return ItemStack.EMPTY;
-            }
-            // Paper end
             if (stack.isEmpty()) {
                 return new ItemStack(Items.GLASS_BOTTLE);
             }
diff --git a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
index 5395dbf1c9bedc1e3e13f6661affdfd94fded502..c31b1411a4fa3f5c474b99c9f032cd255993deff 100644
--- a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
+++ b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
@@ -874,7 +874,6 @@ public class CraftEventFactory {
             if (stack == null || stack.getType() == Material.AIR || stack.getAmount() == 0) continue;
 
             world.dropItem(entity.getLocation(), stack); // Paper - note: dropItem already clones due to this being bukkit -> NMS
-            if (stack instanceof CraftItemStack) stack.setAmount(0); // Paper - destroy this item - if this ever leaks due to game bugs, ensure it doesn't dupe, but don't nuke bukkit stacks of manually added items
         }
 
         return event;
